### PR TITLE
Log reindexing failures

### PR DESCRIPTION
**Pull request by ankikuma**
Log number of reindex tasks in flight during a shutdown. 

This PR aims to address ES-8850. 

Also fixing a small labeling mistake for the reindex metrics.

### CODE REVIEW

**Comment by ankikuma** [comment]
A couple of things to consider while reviewing:
1. Right now I have set the timeout for the wait for reindex tasks to 1 min. Do we want a setting for this so we can tweak it ?
2. Right now I am exposing reindexing failures by logging a message. Do we want to use APM metrics instead ?

File: server/src/main/java/org/elasticsearch/node/Node.java [hunk 1]
```diff
@@ -702,6 +704,52 @@ private void awaitSearchTasksComplete(TimeValue asyncSearchTimeout) {
         }
     }
 
+    private void awaitReindexTasksComplete(TimeValue asyncReindexTimeout) {
```
Comment (Tim-Brooks) [inline server/src/main/java/org/elasticsearch/node/Node.java:707]: Can we combine some code here? Something like:

```java
    private void awaitSearchTasksComplete(TimeValue asyncSearchTimeout) {
        waitForTasks(asyncSearchTimeout, TransportSearchAction.NAME);
    }

    private void awaitReindexTasksComplete(TimeValue reindexTimeout) {
        waitForTasks(reindexTimeout, ReindexAction.NAME);
    }

    private void waitForTasks(TimeValue timeout, String taskName) {
        TaskManager taskManager = injector.getInstance(TransportService.class).getTaskManager();
        long millisWaited = 0;
        while (true) {
            long searchTasksRemaining = taskManager.getTasks().values().stream().filter(task -> taskName.equals(task.getAction())).count();
            if (searchTasksRemaining == 0) {
                logger.debug("all search tasks complete");
                return;
            } else {
                // Let the system work on those searches for a while. We're on a dedicated thread to manage app shutdown, so we
                // literally just want to wait and not take up resources on this thread for now. Poll period chosen to allow short
                // response times, but checking the tasks list is relatively expensive, and we don't want to waste CPU time we could
                // be spending on finishing those searches.
                final TimeValue pollPeriod = TimeValue.timeValueMillis(500);
                millisWaited += pollPeriod.millis();
                if (TimeValue.ZERO.equals(timeout) == false && millisWaited >= timeout.millis()) {
                    logger.warn(
                        format(
                            "timed out after waiting [%s] for [%d] " + taskName + " tasks to finish",
                            timeout.toString(),
                            searchTasksRemaining
                        )
                    );
                    return;
                }
                logger.debug(
                    format("waiting for [%s] " + taskName + " tasks to finish, next poll in [%s]", searchTasksRemaining, pollPeriod)
                );
                try {
                    Thread.sleep(pollPeriod.millis());
                } catch (InterruptedException ex) {
                    logger.warn(
                        format(
                            "interrupted while waiting [%s] for [%d] " + taskName + " tasks to finish",
                            timeout.toString(),
                            searchTasksRemaining
                        )
                    );
                    return;
                }
            }
        }
    }
```

**Review by Tim-Brooks — COMMENTED** [review]
For a test I would create an `IT` performing reindex on a specific node. And then wait for the task to appear in the tasks list. And then immediately stop the node. And check that the task still completes.

You would want to try it out without your shutdown wait code here and make sure you are indexing enough data that the test would (at least sometimes) fail without your shutdown wait you are adding here.

**Comment by Tim-Brooks** [comment]
If it is difficult to get the reindex task to fail from restarting the node, you can always reach in and grab `ThreadPool` and block some of the threads. But I would start by trying to do the test without that.

**Comment by ankikuma** [comment]
Thanks Tim, I will work on a test case.

File: server/src/main/java/org/elasticsearch/node/Node.java [hunk 2]
```diff
@@ -629,6 +630,7 @@ boolean isIncomplete() {
 
             stopperRunner.accept("http-server-transport-stop", injector.getInstance(HttpServerTransport.class)::close);
             stopperRunner.accept("async-search-stop", () -> awaitSearchTasksComplete(maxTimeout));
+            stopperRunner.accept("reindex-stop", () -> awaitReindexTasksComplete(TimeValue.min(TimeValue.timeValueMinutes(1), maxTimeout)));
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/node/Node.java:633]: Let us make the timeout configurable separately from the max-timeout. I think 1 min could be too long by default, I'd go with something like 10secs and we can increase where we need it. The problem is that users do rolling restarts and this could prevent those from completing as fast as before - and the longer it takes to restart the more operations may need to be replayed to re-establish sync'ed replicas.

Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/node/Node.java:633]: `maxTimeout=ZERO` seems to mean infinite, I wonder if the `TimeValue.min` here should respect that?

**Review by henningandersen — COMMENTED** [review]
Looks good, left a couple comments.

**Comment by ankikuma** [comment]
I addressed the review comments and added a test. But I need some help with the test. The reindexing goes through inspite of the shutdown.

In fact, I tried shutting down all 3 nodes in my test - master only, coordinator only and data only. And the reindexing task still goes through. 

I even tried submitting the reindexing task after shutting down the 3 nodes and it still goes through!! I feel like I am missing some very basic understanding of how the integration tests work.

**Comment by ankikuma** [comment]
Added a test. Please take a look @henningandersen @Tim-Brooks whenever you get a chance.

I set the default timeout for the reindexing wait to be 10s. Please let me know if you think I should shorten it.

File: server/src/main/java/org/elasticsearch/node/ShutdownFenceService.java [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+import java.util.function.BiConsumer;
+import java.util.function.Supplier;
+import java.util.stream.Collectors;
+
+import static org.elasticsearch.core.Strings.format;
+
+/**
+ * Extracted out the logic from {@link Node#prepareForClose()} to facilitate testing.
+ */
+public class ShutdownFenceService {
```
Comment (Tim-Brooks) [inline server/src/main/java/org/elasticsearch/node/ShutdownFenceService.java:38]: I know I was the one who created this name initially 🙃, but I'm not sure it is the best. Maybe lets change it to `ShutdownPreparationService`

**Review by Tim-Brooks — COMMENTED** [review]
Mostly looks good to me. Just 1 comment.

File: server/src/main/java/org/elasticsearch/node/ShutdownFenceService.java (hunk 3 above)
Comment (ankikuma) [reply server/src/main/java/org/elasticsearch/node/ShutdownFenceService.java:38]: Sounds good :-)

**Review by Tim-Brooks — APPROVED** [review]
LGTM

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java [hunk 4]
```diff
@@ -0,0 +1,149 @@
...
+            @Override
+            public void onFailure(Exception e) {
+                logger.debug("Encounterd " + e.toString());
+            }
+        };
+        internalCluster().client(coordNodeName).execute(ReindexAction.INSTANCE, reindexRequest, reindexListener);
+
+        // Check for reindex task to appear in the tasks list and Immediately stop coordinating node
+        TaskInfo mainTask = findTask(ReindexAction.INSTANCE.name(), coordNodeName);
+        shutdownPrepareService.prepareForShutdown(taskManager);
```
Comment (henningandersen) [inline modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:130]: I wonder why we need this, should the test not rely on the node stop doing this instead?

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java [hunk 5]
```diff
@@ -0,0 +1,149 @@
...
+        internalCluster().stopNode(coordNodeName);
+    }
+
+    private static TaskInfo findTask(String actionName, String nodeName) {
+        ListTasksResponse tasks;
+        long start = System.nanoTime();
+        do {
```
Comment (henningandersen) [inline modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:137]: Can we use assertBusy() instead?

File: server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java [hunk 6]
```diff
@@ -0,0 +1,167 @@
...
+                        logger.warn("unexpected exception in shutdown task [" + stopper.name() + "]", ex);
+                    } finally {
+                        stopper.listener().onResponse(null);
+                    }
+                }, stopper.name()).start();
+            };
+
+            stopperRunner.accept("http-server-transport-stop", httpServerTransport::close);
+            stopperRunner.accept("async-search-stop", () -> awaitSearchTasksComplete(maxTimeout, taskManager));
+            stopperRunner.accept("reindex-stop", () -> awaitReindexTasksComplete(reindexTimeout, taskManager));
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java:98]: Is this the only change in this class? The refactor makes it a bit harder to reason about the change. We can keep it as is, but it might be beneficial to do the refactor separately for future PRs.

**Review by henningandersen — COMMENTED** [review]
Looks good, left a few comments.

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java [hunk 7]
```diff
@@ -0,0 +1,149 @@
...
+            IntStream.range(0, numDocs)
+                .mapToObj(i -> prepareIndex(INDEX).setId(String.valueOf(i)).setSource("n", i))
+                .collect(Collectors.toList())
+        );
+
+        // Checks that the all documents have been indexed and correctly counted
+        assertHitCount(prepareSearch(INDEX).setSize(0).setTrackTotalHits(true), numDocs);
+    }
+
+    private void checkDestinationIndex(String dataNodeName, int numDocs) {
```
[the remark this answers is not in the document]
Comment (ankikuma) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:77]: Thanks. Done.

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java (hunk 4 above)
Comment (ankikuma) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:130]: Hi Henning, the stopNode() request does not go through the prepareForClose() codepath. I actually tried a "put shutdown" request first to try to exercise my code (without success) until Tim pointed out it doesn't go through prepareForClose(). That is the reason we had to refactor the logic from prepareForClose() into a separate service so we could test it from the IT.

There is this comment in prepareForClose():
"Note that this class is part of infrastructure to react to signals from the operating system - most graceful shutdown logic should use Node Shutdown, see {@link org.elasticsearch.cluster.metadata.NodesShutdownMetadata}."

Since prepareForClose() is triggered by OS signals, I think it is not called in the integration testing framework.

But please advice if you think that prepareForClose() will not be invoked from graceful shutdown codepaths. I do see that "forcefully" calling prepareForClose() is not the ideal way to test that we wait for reindexing tasks during shutdown.

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java [hunk 8]
```diff
@@ -0,0 +1,149 @@
...
+        ensureStableCluster(3);
+
+        int numDocs = 20000;
+        createIndex(numDocs);
+        createReindexTaskAndShutdown(coordNodeName);
+        checkDestinationIndex(dataNodeName, numDocs);
+    }
+
+    private void createReindexTaskAndShutdown(final String coordNodeName) throws Exception {
+
```
[the remark this answers is not in the document]
Comment (ankikuma) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:106]: Done.

File: server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java (hunk 6 above)
Comment (ankikuma) [reply server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java:98]: The refactor was done for testing purpose as explained above.

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java (hunk 5 above)
Comment (ankikuma) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:137]: Sure.

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java [hunk 9]
```diff
@@ -0,0 +1,149 @@
...
+        ActionListener<BulkByScrollResponse> reindexListener = new ActionListener<BulkByScrollResponse>() {
+            @Override
+            public void onResponse(BulkByScrollResponse bulkByScrollResponse) {
+                assertNull(bulkByScrollResponse.getReasonCancelled());
+                logger.debug(bulkByScrollResponse.toString());
+            }
+
+            @Override
+            public void onFailure(Exception e) {
+                logger.debug("Encounterd " + e.toString());
```
[the remark this answers is not in the document]
Comment (ankikuma) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:123]: Sure. I added an assertion error.

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java (hunk 4 above)
Comment (henningandersen) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:130]: Thanks, this makes sense to me, sorry for the noise.

File: server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java (hunk 6 above)
Comment (henningandersen) [reply server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java:98]: Yeah, I understand that. Still, is the line added (plus the method called) the only change here or has other changes sneaked in?

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java (hunk 8 above)
[the remark this answers is not in the document]
Comment (henningandersen) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:106]: I still see it here?

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java (hunk 9 above)
[the remark this answers is not in the document]
Comment (henningandersen) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:123]: Did you not push that yet?

We should probably use `fail(e)` instead (which is almost the same but shorter).

**Review by henningandersen — COMMENTED** [review]
Left a few comments before realizing the changes were not pushed yet.

File: server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java (hunk 6 above)
Comment (ankikuma) [reply server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java:98]: Oh I see, I thought your question was about creating this new class. I only really added the awaitReindexTasksComplete() but in order to reuse the code being used by awaitSearchTasksComplete I created a new API waitForTasks() to replace [awaitSearchTasksComplete()](https://github.com/elastic/elasticsearch/blob/98d53352160d6d2c397ea74567ebc690b701973f/server/src/main/java/org/elasticsearch/node/Node.java#L664)

**Comment by ankikuma** [comment]
I just pushed the changes. I wanted to wait to see if you would suggest a major revamp of the test :)

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java [hunk 10]
```diff
@@ -0,0 +1,145 @@
...
+        flushAndRefresh(DEST_INDEX);
+        assertTrue("Number of documents in source and dest indexes does not match", waitUntil(() -> {
+            final TotalHits totalHits = SearchResponseUtils.getTotalHits(
+                client(dataNodeName).prepareSearch(DEST_INDEX).setSize(0).setTrackTotalHits(true)
+            );
+            return totalHits.relation() == TotalHits.Relation.EQUAL_TO && totalHits.value() == numDocs;
+        }, 10, TimeUnit.SECONDS));
+    }
+
+    private static void findTask(String actionName, String nodeName) throws Exception {
```
Comment (henningandersen) [inline modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:135]: nit: I'd call this waitForTask instead.

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java [hunk 11]
```diff
@@ -0,0 +1,145 @@
...
+    // Make sure all documents from the source index have been reindexed into the destination index
+    private void checkDestinationIndex(String dataNodeName, int numDocs) {
+        assertTrue(indexExists(DEST_INDEX));
+        flushAndRefresh(DEST_INDEX);
+        assertTrue("Number of documents in source and dest indexes does not match", waitUntil(() -> {
+            final TotalHits totalHits = SearchResponseUtils.getTotalHits(
+                client(dataNodeName).prepareSearch(DEST_INDEX).setSize(0).setTrackTotalHits(true)
+            );
+            return totalHits.relation() == TotalHits.Relation.EQUAL_TO && totalHits.value() == numDocs;
+        }, 10, TimeUnit.SECONDS));
```
Comment (henningandersen) [inline modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:132]: I think this can be something like following instead:

```
assertBusy(() -> assertHitCount(prepareSearch(DEST_INDEX).setSize(0).setTrackTotalHits(true), numDocs));
```

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java [hunk 12]
```diff
@@ -0,0 +1,145 @@
...
+    }
+
+    private static void findTask(String actionName, String nodeName) throws Exception {
+        assertBusy(() -> {
+            ListTasksResponse tasks = clusterAdmin().prepareListTasks(nodeName).setActions(actionName).setDetailed(true).get();
+            tasks.rethrowFailures("Find my task");
+            for (TaskInfo taskInfo : tasks.getTasks()) {
+                // Skip tasks with a parent because those are children of the task we want
+                assertFalse(taskInfo.parentTaskId().isSet());
+            }
```
Comment (henningandersen) [inline modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:142]: I think this does not really wait. As a sidenote I am unsure if we need to wait and also whether skipping child tasks is important. But I think this should be:

```suggestion
            for (TaskInfo taskInfo : tasks.getTasks()) {
                // Skip tasks with a parent because those are children of the task we want
                if(taskInfo.parentTaskId().isSet()) == false) {
                    return;
                }
                fail("not found - yet");
            }
```

File: modules/reindex/src/main/java/org/elasticsearch/reindex/ReindexMetrics.java [hunk 13]
```diff
@@ -19,7 +19,7 @@ public class ReindexMetrics {
     private final LongHistogram reindexTimeSecsHistogram;
 
     public ReindexMetrics(MeterRegistry meterRegistry) {
-        this(meterRegistry.registerLongHistogram(REINDEX_TIME_HISTOGRAM, "Time to reindex by search", "millis"));
+        this(meterRegistry.registerLongHistogram(REINDEX_TIME_HISTOGRAM, "Time to reindex by search", "seconds"));
```
Comment (henningandersen) [inline modules/reindex/src/main/java/org/elasticsearch/reindex/ReindexMetrics.java:22]: While I think I agree to the change, I'd prefer to do this in a separate PR.

File: server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java [hunk 14]
```diff
@@ -0,0 +1,171 @@
...
+ * <p>
+ * Note that this class is part of infrastructure to react to signals from the operating system - most graceful shutdown
+ * logic should use Node Shutdown, see {@link org.elasticsearch.cluster.metadata.NodesShutdownMetadata}.
+ */
+public class ShutdownPrepareService {
+
+    private final Logger logger = LogManager.getLogger(ShutdownPrepareService.class);
+    Settings settings;
+    HttpServerTransport httpServerTransport;
+    TerminationHandler terminationHandler;
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java:49]: We should make all fields as final and private as possible. I think these can be private final?

File: server/src/main/java/org/elasticsearch/node/Node.java [hunk 15]
```diff
@@ -167,6 +156,12 @@ public class Node implements Closeable {
         Setting.Property.NodeScope
     );
 
+    public static final Setting<TimeValue> MAXIMUM_REINDEXING_TIMEOUT_SETTING = Setting.positiveTimeSetting(
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/node/Node.java:159]: I think we might as well move these two settings closer to their use, i.e., to `ShutdownPrepareService`.

File: server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+            logger.warn("failed during graceful shutdown tasks", e);
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            logger.warn("interrupted while waiting for graceful shutdown tasks: " + incompleteStoppersDescriber.get(), e);
+        } catch (TimeoutException e) {
+            logger.warn("timed out while waiting for graceful shutdown tasks: " + incompleteStoppersDescriber.get());
+        }
+    }
+
+    private void waitForTasks(TimeValue timeout, String taskName, TaskManager taskManager) {
```
Comment (henningandersen) [inline server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java:126]: Looks like we have differing naming for similar things in this class, I'd prefer to name this: `awaitTasksComplete`, similar to the two other `await` methods.

**Review by henningandersen — APPROVED** [review]
LGTM.

Left a number of smaller comments, but no need for another round.

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java (hunk 11 above)
Comment (ankikuma) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:132]: Yes. Done.

File: modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java (hunk 12 above)
Comment (ankikuma) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:142]: You are right. Thanks for catching this. But I actually don't like using assertBusy() here because it waits for an increasing amount of time and even the first wait is 1 ms. So I do in fact want to change this back to what I had originally where we wait for 10 nanoseconds at a time.

Comment (ankikuma) [reply modules/reindex/src/internalClusterTest/java/org/elasticsearch/index/reindex/ReindexNodeShutdownIT.java:142]: You are right. Thanks for catching this, I fixed it and tested it out as well (by holding off submission of the reindexing task).

File: server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java (hunk 14 above)
Comment (ankikuma) [reply server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java:49]: Yes. Thanks!

File: server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java (hunk 16 above)
Comment (ankikuma) [reply server/src/main/java/org/elasticsearch/node/ShutdownPrepareService.java:126]: Yes, that's a better name. Thanks.

File: server/src/main/java/org/elasticsearch/node/Node.java (hunk 15 above)
Comment (ankikuma) [reply server/src/main/java/org/elasticsearch/node/Node.java:159]: Sure done.

**Comment by ankikuma** [comment]
Wait for reindexing tasks to finish during shutdown for an amount of time defined by settings. Also log the number of reindexing tasks still in flight after the wait.
